src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java

```diff
@@ -667,6 +697,9 @@ private static boolean isLoopAst(int ast) {
         /** Contains definitions of uninitialized variables. */
         private final Deque<DetailAST> uninitializedVariables = new ArrayDeque<>();
 
+        /** Whether there is a {@code break} or {@code return} in the scope. */
+        private boolean containsBreakOrReturn;
```

Comment:
> The first example is handled by the check just like the one I wrote at #4115 (comment):

Your example had a `break` in a while loop. My example has a `return` instead. It will not produce a compile issue like you were stating.

>  It is no longer just a "return" and "break" problem now.

Let's focus on these specific tokens for now. We can create new issue for others.
If you find supporting `break` and `return` too hard, and if @romani agrees, we can split this into new issues.
I have only seen false negatives so far, not any false positives.